### PR TITLE
Added basic glog functionality keeping compatibility with spdlog

### PR DESCRIPTION
**Pull request by mrice32**
`--cxxopt=-DGLOG_ON` is the bazel flag to turn glog on

### CODE REVIEW

**Comment by mrice32** [comment]
Also, the docker builds and tests won't work because I've added a dependency.  I hacked together something locally to make it work (building a new image and pointing the script at it), but we'll have to ultimately update the remote docker image to get it working.

File: source/common/common/logger.h [hunk 1]
```diff
@@ -123,48 +195,16 @@ template <Id id> class Loggable {
...
-#define stream_log_trace(FORMAT, STREAM, ...)                                                      \
-  stream_log(log(), trace, FORMAT, STREAM, ##__VA_ARGS__)
-#define stream_log_debug(FORMAT, STREAM, ...)                                                      \
-  stream_log(log(), debug, FORMAT, STREAM, ##__VA_ARGS__)
-#endif
-
-#define stream_log_info(FORMAT, STREAM, ...) stream_log(log(), info, FORMAT, STREAM, ##__VA_ARGS__)
+// Function for formatting the connection log
+template <typename T, typename... Args>
+std::string format_connection_log(const std::string& format, T& connection, Args&&... args) {
```
Comment (mrice32) [inline source/common/common/logger.h]: I'm not sure if these should be macros or regular functions.  A macro might make it a bit faster by concatenating the two string literals at compile time rather than forcing the creation of two string objects, but that might not be a concern.

Comment (jamessynge) [reply source/common/common/logger.h]: My inclination is to not worry about it until it shows up as a problem.

File: source/common/common/logger.h [hunk 2]
```diff
@@ -123,48 +195,16 @@ template <Id id> class Loggable {
...
-#define stream_log_debug(...)
-#else
-#define stream_log_trace(FORMAT, STREAM, ...)                                                      \
-  stream_log(log(), trace, FORMAT, STREAM, ##__VA_ARGS__)
-#define stream_log_debug(FORMAT, STREAM, ...)                                                      \
-  stream_log(log(), debug, FORMAT, STREAM, ##__VA_ARGS__)
-#endif
-
-#define stream_log_info(FORMAT, STREAM, ...) stream_log(log(), info, FORMAT, STREAM, ##__VA_ARGS__)
+// Function for formatting the connection log
```
Comment (jamessynge) [inline source/common/common/logger.h]: "a connection log." (i.e. the->a, trailing period.)

File: source/common/common/logger.h [hunk 3]
```diff
@@ -123,48 +195,16 @@ template <Id id> class Loggable {
...
+template <typename T, typename... Args>
+std::string format_connection_log(const std::string& format, T& connection, Args&&... args) {
+  return fmt::format("[C{}] " + format, connection.id(), std::forward<Args>(args)...);
+}
+
+// Function for formatting the stream log
```
Comment (jamessynge) [inline source/common/common/logger.h]: "a stream log."

File: source/common/filesystem/watcher_impl.cc [hunk 4]
```diff
@@ -75,7 +77,7 @@ void WatcherImpl::onInotifyEvent() {
 
       for (FileWatch& watch : callback_map_[file_event->wd].watches_) {
         if (watch.file_ == file && (watch.events_ & events)) {
-          log_debug("matched callback: file: {}", file);
+          VLOG(1) << fmt::format("matched callback: file: {}", file);
```
Comment (jamessynge) [inline source/common/filesystem/watcher_impl.cc]: Given that log_debug is compiled out for non-debug builds, this should be DVLOG (here and elsewhere).
The same is true with log_trace also.

File: source/server/config/network/http_connection_manager.cc [hunk 5]
```diff
@@ -150,9 +150,9 @@ HttpConnectionManagerConfig::HttpConnectionManagerConfig(const Json::Object& con
     std::string string_name = filters[i]->getString("name");
     Json::ObjectSharedPtr config_object = filters[i]->getObject("config");
 
-    log().info("    filter #{}", i);
-    log().info("      type: {}", string_type);
-    log().info("      name: {}", string_name);
+    LOG(INFO) << fmt::format("    filter #{}", i);
```
Comment (jamessynge) [inline source/server/config/network/http_connection_manager.cc:153]: Perhaps add a TODO here to restructure this as 1 LOG line.

File: source/server/configuration_impl.cc [hunk 6]
```diff
@@ -166,9 +166,9 @@ MainImpl::ListenerConfig::ListenerConfig(MainImpl& parent, Json::Object& json) :
     std::string string_type = filters[i]->getString("type");
     std::string string_name = filters[i]->getString("name");
     Json::ObjectSharedPtr config = filters[i]->getObject("config");
-    log().info("  filter #{}:", i);
-    log().info("    type: {}", string_type);
-    log().info("    name: {}", string_name);
+    LOG(INFO) << fmt::format("  filter #{}:", i);
```
Comment (jamessynge) [inline source/server/configuration_impl.cc:169]: Ditto.

File: source/server/connection_handler_impl.cc [hunk 7]
```diff
@@ -13,7 +14,9 @@ namespace Envoy {
 namespace Server {
 
 ConnectionHandlerImpl::ConnectionHandlerImpl(spdlog::logger& logger, Api::ApiPtr&& api)
-    : logger_(logger), api_(std::move(api)), dispatcher_(api_->allocateDispatcher()) {}
+    : logger_(logger), api_(std::move(api)), dispatcher_(api_->allocateDispatcher()) {
+  UNREFERENCED_PARAMETER(logger_);
```
Comment (jamessynge) [inline source/server/connection_handler_impl.cc:18]: Why is this needed? To handle the GLOG_ON case, which case I assume that logger_ is an unused member, rather than unused parameter.

File: source/server/drain_manager_impl.cc [hunk 8]
```diff
@@ -50,7 +50,7 @@ void DrainManagerImpl::startParentShutdownSequence() {
   ASSERT(!parent_shutdown_timer_);
   parent_shutdown_timer_ = server_.dispatcher().createTimer([this]() -> void {
     // Shut down the parent now. It should have already been draining.
-    log().warn("shutting down parent after drain");
+    LOG(WARNING) << fmt::format("shutting down parent after drain");
```
Comment (jamessynge) [inline source/server/drain_manager_impl.cc]: Please remove fmt::format where there are no parameters in the log message.

**Review by jamessynge — COMMENTED** [review]
Looking good, almost there.

File: source/server/connection_handler_impl.cc (hunk 7 above)
Comment (mrice32) [reply source/server/connection_handler_impl.cc:18]: Yes, the compiler complains in the GLOG_ON case.  Is there a preferred way of ignoring the member?  We could make another macro for unused members that puts a compiler directive in front of it where it's declared.

File: source/server/config/network/http_connection_manager.cc (hunk 5 above)
Comment (mrice32) [reply source/server/config/network/http_connection_manager.cc:153]: Why not just restructure it as a part of this PR?

File: source/common/filesystem/watcher_impl.cc (hunk 4 above)
Comment (mrice32) [reply source/common/filesystem/watcher_impl.cc]: Nice catch.  That would've been a big behavior change.

**Comment by mrice32** [comment]
@jamessynge I believe I have corrected for all of your comments except for the two that I responded to with questions.

**Comment by mrice32** [comment]
After discussing with Matt, I'm closing this in favor of #3.
